src/Type/Php/RegexArrayShapeMatcher.php

```diff
@@ -154,18 +155,24 @@ private function matchRegex(string $regex, ?int $flags, TrinaryLogic $wasMatched
 			);
 
 			if (!$this->containsUnmatchedAsNull($flags, $matchesAll)) {
+				// positive match has a subject but not any capturing group
 				$combiType = TypeCombinator::union(
 					new ConstantArrayType([new ConstantIntegerType(0)], [$this->createSubjectValueType($flags, $matchesAll)], [0], [], true),
 					$combiType,
 				);
 			}
 
+			$onlyOptionalTopLevelGroup->clearOverrides();
```

Comment:
This smells of too much statefulness and mutability. Is this pattern in more places in Regex\? I'd much prefer immutable stateless value objects.

Reply:
totally agree. 

I tried doing it in a immutable way, but my attempts don't work because the objects beeing changed here, are referenced by other objects via `parent` property (and we want the objects referenced by `$onlyOptionalTopLevelGroup` to change, no matter which object in the `$groupList` references it.

Reply:
> Is this pattern in more places in Regex?

we use this pattern for the 2 tagged union special cases in this method here. nowhere else.